### ISSUE TITLE
timestamp not updating with pvput for PVs on PvaServer

### ISSUE THREAD

**Issue by kathryn-baker** [epics-base/pvaPy#67]
Hi there, I've been using pvapy to set up my own PV server and update the values of the timestamps and PVs through the PvObject set method as a message arrives from another system. So far this works very well and the timestamps update as expected. 

However, I would also like to be able to control the value of PVs through the standard pvput command line tool and I've found that I can do this fine to set the value, but the command doesn't update the timestamp, instead it uses the timestamp declared previously, e.g. 
```
test 1990-01-01 00:00:00.456  -1 0
test 1990-01-01 00:00:00.456  -1 5
```

I've tried fixing this by setting the timestamp manually using the server.update() method within an onWriteCallback assigned to the PV when it is added to the server (see the code below). What this appears to do is update the value, and then immediately after it updates the timestamp, so when I monitor the channel I receive two updates for every pvput, like this:
```
test 1990-01-01 00:00:00.456  -1 0
test 1990-01-01 00:00:00.456  -1 5
test 2021-08-13 15:09:29.659  -1 5
test 2021-08-13 15:09:29.659  -1 6
test 2021-08-13 15:09:38.949  -1 6
test 2021-08-13 15:09:38.949  -1 7
test 2021-08-13 15:09:58.518  -1 7
```
In order to apply some extra formatting to the values from the other system I am using an object that inherits from the PvObject class. I can see that in the onWriteCallback a standard PvObject is passed as the parameter which is why I have to call set on my inherited PvObjects within the callback. I understand this would generate a new update (as a new set method is called), but I'm not sure how to avoid this if I want to manually update the timestamp. 

Is it essential that the object is an ordinary PvObject for the server.update method to work, or have I misunderstood how the pvput command relates to the PvaServer and the onWriteCallback, and that actually it isn't supposed to update the timestamp anyway?

I've included some code below that is a representation of how I am currently using the onWriteCallback and inherited classes to help illustrate what I mean. Running this with pvput produces the pvmonitor output shown above.  

```python
from pvapy import PvaServer, INT, PvObject, STRING, PvTimeStamp
import time

def extract_seconds_and_nanoseconds_from_timestamp(timestamp):
    seconds = int(timestamp // 1)
    nanoseconds = int((timestamp % 1) * 1e9)
    return seconds, nanoseconds


class ServerManager():
    def __init__(self):
        self.pvaserver = PvaServer()
        self.pv = None

    def create_pv_and_add_to_server(self, name):
        timestamp = PvTimeStamp(123, 456000000)
        pv = InheritedPvObject(
            {'channelname': STRING, 'value': INT, 'timeStamp': timestamp},
            {'channelname': name, 'timeStamp': timestamp.toDict()}
        )
        self.pv = pv
        self.pvaserver.addRecord(name, self.pv.copy(), self.echo)

    def echo(self, pv_object):
        # extract the information from the pvobject that comes through (different type to
        # the inherited one so we can't assign it straight off)
        pv_object_fields = pv_object.get()

        # update the timestamp
        seconds, nanoseconds = extract_seconds_and_nanoseconds_from_timestamp(
            time.time())
        timestamp = PvTimeStamp(seconds, nanoseconds)
        pv_object_fields['timeStamp'] = timestamp.toDict()

        # set the new value in the Inherited object
        self.pv.set(pv_object_fields)

        # update the inherited object on the server with the new information
        channel_name = pv_object_fields.get('channelname')
        self.pvaserver.update(
            channel_name, self.pv)


class InheritedPvObject(PvObject):
    def __init__(self, type_dict, value_dict):
        PvObject.__init__(self, type_dict, value_dict)


if __name__ == '__main__':
    try:
        server = ServerManager()
        server.create_pv_and_add_to_server('test')
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print('stopping server')
    finally:
        server.pvaserver.stop()
```

I am currently working in Python 3.6.8 with pvapy (3.1.0) on CentOS Linux release 8.4.2105, alongside EPICS version R7.0.5

Any guidance as to why this issue may be arising would be a great help. Thank you.

**Comment by sveseli** [comment]
Thanks for reporting this. The onWriteCallback you have will post a second update to the record, and hence you see two updates for each pvput. This is working as expected. We did have a problem with record processing, however.
 
I just pushed fix (commit dff962c3) for timestamp updates on the master branch. It works as follows:

Terminal 1 (server):
<pre>
>>> from pvaccess import *
>>> def onWrite(pv):
...     print(pv)
... 
>>> pv = PvObject({'value' : INT, 'timeStamp' : PvTimeStamp()})
>>> s = PvaServer('x', pv, onWrite)
</pre>

Terminal 2 (client):
<pre>
$ pvget x
x <undefined>              0 
$ pvput x 15
Old : <undefined>              0 
New : 2021-08-13 15:54:29.664  15 
$ pvget x
x 2021-08-13 15:54:29.664  15 
</pre>

After pvput, terminal 1 shows the above onWrite callback being invoked:
<pre>
>>> structure 
    int value 15
    time_t timeStamp
        long secondsPastEpoch 1628888069
        int nanoseconds 663647801
        int userTag 0
</pre> 

Will this help?
